src/Processors/QueryPlan/QueryPlan.cpp

```diff
@@ -429,6 +430,65 @@ void QueryPlan::explainPipeline(WriteBuffer & buffer, const ExplainPipelineOptio
     }
 }
 
+void QueryPlan::explainEstimates(MutableColumns & columns)
+{
+    checkInitialized();
+
+    struct Frame
+    {
+        Node * node = nullptr;
+        bool is_description_printed = false;
+        size_t next_child = 0;
+    };
+
+    using CountersPtr = std::shared_ptr<QueryPlan::EstimateCounters>;
+    std::unordered_map<std::string, CountersPtr> counters;
+    std::stack<Frame> stack;
+    stack.push(Frame{.node = root});
+
+    while (!stack.empty())
+    {
+        auto & frame = stack.top();
+
+        if (!frame.is_description_printed)
```

Comment:
This is somewhat confusing. A 'frame' is just a plan node, right? And this is a depth-first traversal. It should visit all the nodes once, why do you need a special flag to ensure this? A recursive function is probably good enough, these plans are not so deep.

Reply:
Ok, we just traverse the plan tree. I will fix it.